Add Implementation Shortfall models with the shortfall() function

vi-to opened the pull request:
The present pull request is a starting point to add the "_Implementation Shortfall_" algorithmic TCA models, firstly issued by @jaymon0703  in issue [#54](https://github.com/braverock/blotter/issues/54), by introducing the function `shortfall`.

The present function is a development version that can (and I am sure it will) be improved in many directions. I have tried to provide a minimal working structure. It first of all needs a correctness check and then the addition of the others "higher" level aspects.

In fact, I have been the more explicit I could to facilitate this phase. For example, I left the several alternative simplified expressions of the IS methods in terms of the average execution price (`p_avg`), which are provided in Kissell's book, as a side to the corresponding extended ones, so that we can then decide whether to consistently implement them along the function.

Your observations and comments are kindly encouraged. As the most immediate thing I can think of, there is the style names of variables and parameters, I have tried to choose names that seemed similar to the style of the ones currently used elsewhere in the package, but I am open to hearing alternatives that would better integrate the function into the package (eventually).
Several other comments and questions will be added in reference to specific lines of code.

vi-to committed:
First commit to add shortfall() function

vi-to committed:
Documentation of shortfall() function

vi-to committed:
Update shortfall function docs

vi-to commented on R/shortfall.R:
```diff
@@ -0,0 +1,129 @@
+#' @param PaQty The total number of units (such as shares or contracts) intended to trade
+#' @param PaPriceStart Decision price at the beginning of trading
+#' @param PaPriceEnd Decision future price at the end of trading
+#' @param method String specifying which method to use for The Implementation Shortfall calculation. One of 'Perold' or 'Wagner'
```
Hi @jaymon0703, I have noticed you updated the issue with a very comprehensive presentation of the IS models. 

There you proposed to implement the function using the parameter `model=c("Market","Perold","Wagner")`. Before this was online I used `method` in a very similar fashion, would it be better to change it? 

Also, you proposed to default this parameter to the Market Activity Implementation Shortfall (Mkt.ActivityIS), but I am not sure I fully understand. Would you please provide further comments on this? 
Currently, the parameter defaults to what Kissell calls the 'Complete execution' case, which is the only case in which the shortfall is calculated assuming that all the intended trade units (`PaQty`) were executed. Strictly speaking, I would not consider it to be a method or model per se and this is why I put it before the "proper" methods/models.
The Mkt.ActivityIS is implicitly implemented as a particular case of Wagner's method (I left a note at line 114). It seems to me the easier way to implement it, its difference is only in the meaning of the output. In a certain sense, it is the default behavior of the function when `method='Wagner'`, because when no `PaPriceStart` is provided, the `ArrPrice` (what you named `p_0`) is used by default. What do you think if I stress this aspect in the documentation or we make it produce a message when this happens?

jaymon0703 replied:
Thanks for the PR @vi-to 

I think a good general architecture can be:
If no decision price is specified, then use Market Activity IS (equation 3.8) with arrPrice as you have specified. The key difference between Market Activity IS and the others is that no decision price is specified. We may want to add arrPrice as a separate parameter, since arrival price is really the middle of the bid-ask spread at the time the order enters the orderbook. If this is not specified, then use first transaction price. Also, Market Activity IS is exactly Expanded (ie. Wagner) IS without the Delay Cost (ie. first term S(P0-Pd)).

If decision price is present, then the other 2 methods (Wagner & Perold) come into play.

I think method=c("Market", "Perold", "Wagner") would be useful, and we should make the methods explicit.

Lastly, it may be worth including an example, such as 

```
data(amzn)
txns <- amzn.trades[which(amzn.trades[,2] > 0),]
try(rm("portfolio.testport","account.testport",pos=.blotter), silent = TRUE)
stock.str='amzn' # what are we trying it on
currency('USD')
stock(stock.str,currency='USD',multiplier=1)
initPortf("testport", symbols=stock.str)
initAcct("testport","testport", symbols=stock.str)
addtxns <- addTxns("testport","amzn",txns)
updatePortf("testport")
p = getPortfolio("testport") # make a local copy of the portfolio object
a = getAccount("testport") # make a local copy of the account object
p$symbols$amzn$txn

implementationShortfall("testport","amzn",3300,127,128, method = 'Market')
implementationShortfall("testport","amzn",3300,127,128,method = 'Perold')
implementationShortfall("testport","amzn",3300,127,128,method = 'Wagner')
```

The respective Implementation Shortfall metrics for Perold and Wagner should be identical when used with similar inputs, so i would double check your price used for ExeCost in the last term for Perold, and also your calc for TradeCost in Wagner. Fix those and you should get measures which reconcile when used with identical inputs.

vi-to replied:
Found two errors in current formulations (ironically alternatives were 'correct'), committing the corrections.
Next, I am trying to move to this architecture, but would like to give it some additional thoughts, I will submit related commits to your attention later in the day.

jaymon0703 replied:
Perfect, thanks @vi-to

vi-to replied:
Hi @jaymon0703, the last commit adds a new flow for methods where `method='Market'` is explicitly accounted for. The default `method` has not changed yet, as I feel that it may be useful to have for "quick calcs". However, we could output `PaRet` and `AcRet` elsewhere or, perhaps a better option, is to make it a "method". What do you think about this?

Also, I would like to suggest you a test (which is related to the conversation we are having about the defaults values of the other parameters) you might want to add to the ones you are running while reviewing. When `PaPriceStart = ArrPrice`, the methods must of course return the same shortfall measure (up to approxs.); for example, using the last data you provided, here's my output:

```
> shortfall("testport", "amzn", PaQty=3300, PaPriceStart=128, PaPriceEnd=130, ArrPrice=128, method='Perold')
  Symbol Txn.Qty u.Txn.Qty Exe.Cost Opp.Cost Fees Shortfall
1   amzn    3300         0    -2830        0    0     -2830
> shortfall("testport", "amzn", PaQty=3300, PaPriceStart=128, PaPriceEnd=130, ArrPrice=128, method='Wagner')
  Symbol Txn.Qty u.Txn.Qty Opp.Delay Trade.Delay Delay.Cost Trade.Cost Opp.Cost Fees Shortfall
1   amzn    3300         0         0           0          0      -2830        0    0     -2830
> shortfall("testport", "amzn", PaQty=3300, PaPriceStart=128, PaPriceEnd=130, ArrPrice=128, method='Market')
  Symbol Txn.Qty u.Txn.Qty        Trade.Cost Opp.Cost Fees         Shortfall
1   amzn    3300         0 -2830.00000000001        0    0 -2830.00000000001
```

jaymon0703 replied:
Thanks Vito

I have had a chance to review the PR and have a few observations and suggestions.

1. It would be useful to specify all the methods in the functions signature, such as method = c("Complete", "Market", "Perold", "Wagner"). If no value is specified, the default will be Complete. I agree, having Complete as an available method will be useful.

2. This speaks to your question in issue #81 and i think its worth addressing sooner rather than later in case others are looking for guidelines for contributing to blotter. Happy to work with you on a contributing guidelines doc along the lines of https://help.github.com/en/articles/setting-guidelines-for-repository-contributors. Nevertheless, we tend to have a preference for camelCase when setting variable or param names. So for example ArrPrice could rather be arrPrice and PaPriceStart could rather be priceStart, etc.

3. Your calculation for Complete Execution looks incorrect. Pn is not relevant when all trades intended for execution are ultimately executed. I have updated the issue #54 with the formula we should be using for Complete Execution. In AcRet replace PaPriceEnd with the average execution price. Paper return is irrelevant, because the entire quantity was traded.

4. Fees are added as negative amounts in AddTxns, so we will have to subtract the fees variable in order to add it to the Implementation Shortfall measure. So in line 89, perhaps `Fees <- -1 * sum(txns$Txn.Fees)`

5. Can we rename the file impShortfall and likewise the function name? Shortfall can be confused with other functions in other packages such as Expected Shortfall in PerformanceAnalytics so we should be a bit more explicit there. In the long run we may want to consolidate all TCA models into one TCA.R file, but for now impShortfall.R will do.

6. I think it would be useful to add a column to the dataframe output for the type of method used. For example, if someone loops through all methods they would want to see which method relates to which output.

7. We should think about ways to gracefully handle calls to `impShortfall` when certain params are missing. For example, for Complete Execution the price at the end of the trading horizon is irrelevant, so users should not have to specify a value for that if using method="Complete". For method="Market" the decision price is not required, so perhaps specifying PaPriceStart=ArrPrice for example will let us call shortfall() without the user having to specify a decision price.

8. I have set up dummy trades to mirror the examples in the book and i think these could be useful examples in the documentation...once the change for fees and the correction for "Complete" are implemented i think we will be in a position to merge with the tca branch in blotter. If you dont mind i would like to add a bit more context in the documentation, once it is merged.

Thanks again for the PR.

Tests to replicate the examples in the book...

```
# set up test_txns assuming all 5k shares traded
test_txns <- xts(cbind(rep(10:11,5),rep(500,10),TxnFees = rep(-10,10)),
                 seq.POSIXt(as.POSIXct("2000-01-01 09:00:00"), as.POSIXct("2000-01-01 18:00:00"), by = "hours"))
colnames(test_txns) <- c("TxnPrice","TxnQty","TxnFees")

stock.str='tca_test' # what are we trying it on
currency('USD')
stock(stock.str,currency='USD',multiplier=1)
suppressWarnings(rm("account.testport","portfolio.testport",pos=.blotter))
initPortf("testport", symbols=stock.str)
initAcct("testport","testport", symbols=stock.str)
addtxns <- addTxns("testport","tca_test",test_txns)
updatePortf("testport")
p = getPortfolio("testport") # make a local copy of the portfolio object
a = getAccount("testport") # make a local copy of the account object
p$symbols$tca_test$txn

### Complete Execution
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice=10)

#### Market
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice = 10, method='Market')
# If change in point 7 is implemented, then the call to shortfall can be:
shortfall("testport", "tca_test", PaQty=5000, PaPriceEnd=11, ArrPrice = 10, method='Market')

# set up test_txns assuming assuming only 4k shares traded
test_txns <- test_txns[-c(2:3),]
suppressWarnings(rm("account.testport","portfolio.testport",pos=.blotter))
initPortf("testport", symbols=stock.str)
initAcct("testport","testport", symbols=stock.str)
addtxns <- addTxns("testport","tca_test",test_txns)
updatePortf("testport")
p = getPortfolio("testport") # make a local copy of the portfolio object
a = getAccount("testport") # make a local copy of the account object
p$symbols$tca_test$txn

### Perold
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice=10, method='Perold')

### Wagner
shortfall("testport", "tca_test", PaQty=5000, PaPriceStart=10, PaPriceEnd=11, ArrPrice=10.25, method='Wagner')
```

vi-to replied:
Thank you @jaymon0703 for your comprehensive review.

I will first address (1), (4) and (6). 

Then (7), (2), (5) and docs update also with examples in (8), possibly in this order. 

As for (3):
while the alternative expression is correct, I chose to implement it fully to be consistent with the others (at the moment the function is completely independent of `p_avg`) and provide its two underneath quantities. When the provided `PaQty` equals `sum(txns$Txn.Qty)`, they are equivalent. It may happen however that one provides an inconsistent `PaQty`. Furthermore, once `PaPriceEnd` default value will be in place, there will be no need to provide it, especially when `method='Complete'` (in accordance with (7)).
If you think the full implementation is valuable, here I propose to force `PaQty` to `sum(txns$Txn.Qty)` and add a warning, or even make this by default with a missing `PaQty` since it is the purpose of the method.
Taking it further, this can be a default of the function as it would allow using other methods as well, with obvious zero opportunity cost returns and other shortfall components adapted to executed trades (and other params).

About (8):
what you are proposing is great, indeed in doing so I encourage you to add yourself as an author of the function if you want to.
Related with the docs is that, once we have finished the implementation, I was thinking to factor out `sum(txns$Txn.Qty)` to simply the expressions and rename it to avoid possible confusions with other `blotter` produced outputs (~~maybe `TotTxnQty`~~ `tTxnQty`).

jaymon0703 replied:
Ok thanks @vi-to, looks like your latest commit resolves 1, 4 and 6. Once 7, 2, 5 and 8 are resolved we can merge this and move on.

vi-to replied:
@jaymon0703 I am about to commit the function name change. I have corrected spacing as well, at this point do you want me to change the `.R` name too?

vi-to commented on R/shortfall.R:
```diff
@@ -0,0 +1,129 @@
+#' @param Symbol A string identifying the symbol to calculate the Shortfall for
+#' @param PaQty The total number of units (such as shares or contracts) intended to trade
+#' @param PaPriceStart Decision price at the beginning of trading
+#' @param PaPriceEnd Decision future price at the end of trading
```
@jaymon0703, P_n is made available as `PaPriceEnd` (which now I notice should be better explained in the context of the documentation I wrote above), but has no default value. You suggested defaults to the closing price for the day or the price of the last executed transaction. While I haven't a solution for the former yet, how can we go for the latter? `last(txns$Txn.Price)`?

jaymon0703 replied:
Forgive me, the suitable default will be 'last(txns$Txn.Price)' with the ideal value being the mid-price at the end of the trading horizon. All prices, during market hours, are meant to be at the mid-price. If final transaction is in a close auction, then the transaction price will be at the close price for the day anyway. I have updated that section of the issue.

vi-to replied:
In addition to `PaPriceEnd`, for which we do have a default value in mind, at the moment `PaPriceStart` hasn't one. In the issue you proposed to use the arrival price, but now `ArrPrice` is a param. Should we similarly go with `first(txns$Txn.Price)` for this last one and then just set their 'default relation'?

vi-to committed:
shortfall: correct two errors in methods formulations

vi-to committed:
shortfall: add major change in methods flow, add 'ArrPrice' as a new param, update docs

vi-to committed:
Add new method 'Complete' as default of 'method' argument, correct 'Fees' value and update docs

vi-to committed:
Correct 'Market' method return values docs

vi-to committed:
Add params default values, add a warning message and update docs

Params 'PaQty', 'PaPriceEnd', 'ArrPrice' and 'PaPriceStart' have now default values.
When param method='Complete' is used inconsistently, it will be corrected and a warning will be produced

vi-to committed:
Change params names and convert them to camelCase style

Params that changed and their changes:
PaQty is now paQty
PaPriceStart is now priceStart
PaPriceEnd is now priceEnd
ArrPrice is now arrPrice

Variables changed as well. Check diff.

vi-to committed:
Simplify expressions introducing tTxnQty variable and update docs
